cdap-watchdog/src/main/java/co/cask/cdap/logging/pipeline/logbuffer/LogBufferProcessorPipeline.java

```diff
@@ -0,0 +1,361 @@
+      }
+    }
+  }
+
+  /**
+   * Persists the checkpoints.
+   */
+  private void persistCheckpoints() {
+    try {
+      checkpointManager.saveCheckpoints(checkpoints);
+      LOG.debug("Checkpoint persisted for {} with {}", name, checkpoints);
+    } catch (Exception e) {
+      // Just log as it is non-fatal if failed to save checkpoints
+      OUTAGE_LOG.warn("Failed to persist checkpoint for pipeline {}.", name, e);
+    }
+  }
+
+  /**
+   * Sync the appender and persists checkpoints if it is time.
+   *
+   * @return delay in millisecond till the next sync time.
+   */
+  private long trySyncAndPersistCheckpoints(long currentTimeMillis) {
+    if (unSyncedEvents == 0) {
+      return config.getCheckpointIntervalMillis();
+    }
+    if (currentTimeMillis - config.getCheckpointIntervalMillis() < lastCheckpointTime) {
+      return config.getCheckpointIntervalMillis() - currentTimeMillis + lastCheckpointTime;
+    }
+
+    // Sync the appender and persists checkpoints
+    try {
+      context.sync();
+      // Only persist if sync succeeded. Since persistCheckpoints never throw, it's ok to be inside the try.
+      persistCheckpoints();
+      lastCheckpointTime = currentTimeMillis;
+      metricsContext.gauge("last.checkpoint.time", lastCheckpointTime);
+      unSyncedEvents = 0;
+      LOG.debug("Events synced and checkpoint persisted for {}", name);
+    } catch (Exception e) {
+      OUTAGE_LOG.warn("Failed to sync in pipeline {}. Will be retried.", name, e);
+    }
+    return config.getCheckpointIntervalMillis();
+  }
+
+  /**
+   * A mutable implementation of {@link Checkpoint}.
+   */
+  private static final class MutableLogBufferCheckpoint extends Checkpoint<LogBufferFileOffset> {
+    private MutableLogBufferFileOffset offset;
+    private long maxEventTs;
+
+    MutableLogBufferCheckpoint(long fileId, long filePos, long maxEventTs) {
+      super(new LogBufferFileOffset(fileId, filePos), maxEventTs);
+      this.offset = new MutableLogBufferFileOffset(fileId, filePos);
+      this.maxEventTs = maxEventTs;
+    }
+
+    @Override
+    public MutableLogBufferFileOffset getOffset() {
+      return offset;
+    }
+
+    @Override
+    public long getMaxEventTime() {
+      return maxEventTs;
+    }
+
+    void setMaxEventTs(long maxEventTs) {
+      this.maxEventTs = maxEventTs;
+    }
+
+    @Override
+    public String toString() {
+      return "MutableLogBufferCheckpoint{" +
+        "offset=" + offset +
+        ", maxEventTs=" + maxEventTs +
+        '}';
+    }
+  }
+
+  /**
+   * A mutable implementation of {@link LogBufferFileOffset}.
+   */
+  private static final class MutableLogBufferFileOffset extends LogBufferFileOffset {
+    private long fileId;
+    private long filePos;
+
+    MutableLogBufferFileOffset(long fileId, long filePos) {
+      super(fileId, filePos);
+      this.fileId = fileId;
+      this.filePos = filePos;
+    }
+
+    @Override
+    public long getFileId() {
```

Comment:
if everything here is private, can access the fields directly rather than through getters and setters.

Reply:
This is for checkpoint manager to access mutated state.

Reply:
oh I see, these are overrides.